README.md
# hichihikers-haskell
 stack new hichihikers-haskell
 cd hichihikers-haskell
 stack exec -- hichihikers-haskell-exe
 stack build --fast --file-watch --exec hichhiker-haskell-exe
 stack build --fast --file-watch --exec hichhikers-haskell-exe
 stack exec -- hichihikers-haskell-exe
 stack build --fast --file-watch --exec hichhikers-haskell-exe
 stack build --fast --file-watch --exec ./hichhikers-haskell-exe
 stack build --fast --file-watch --exec hichhikers-haskell-exe
 stack exec -- hichihikers-haskell-exe
 stack build --fast --file-watch --exec hichihikers-haskell-exe
 du -s * | runshaskell ./src/Lib2.hs
 du -s * | runhaskell ./src/Lib2.hs
 echo "foo" | runhaskell ./src/Lib2.hs
 du -s * | runhaskell ./src/Lib2.hs
 sb -s * | runhaskell ./src/Lib2.hs
 du -s * | runhaskell ./src/Lib2.hs

 du -s * | stack exec -- hichihikers-haskell-exe


ChangeLog.md
# Changelog for hichihikers-haskell

## Unreleased changes


test/Spec.hs
import Control.Monad (liftM2, replicateM)
import Lib2
import Test.QuickCheck

instance Arbitrary Dir where
  arbitrary = liftM2 Dir gen_size gen_name
    where
      gen_size = do
        s <- choose (10, 14000)
        return (s * 1024 * 1024)
      gen_name = do
        n <- choose (1, 300)
        replicateM n (elements "fubar/")

prop_greedy_pack_is_fixpoint ds =
  let pack = greedy_pack ds
   in pack_size pack == pack_size (greedy_pack (dirs pack))

main :: IO ()
main = do
  putStrLn "Lib2 test started -----"
  quickCheck prop_greedy_pack_is_fixpoint


test/unit/Spec.hs
import Control.Monad (liftM2, replicateM)
import Lib3
import Test.QuickCheck

instance Arbitrary Dir where
  arbitrary = liftM2 Dir gen_size gen_name
    where
      gen_size = do
        s <- choose (10, 14000)
        return (s * 1024 * 1024)
      gen_name = do
        n <- choose (1, 300)
        replicateM n (elements "fubar/")

--prop_dynamic_pack_is_fixpoint ds =
--let pack = dynamicPack ds
--in pack_size pack == pack_size (dynamicPack (dirs pack))

prop_dynamic_pack_small_disk ds =
  let pack = dynamicPack 50000 ds
   in pack_size pack == pack_size (dynamicPack 50000 (dirs pack))

main :: IO ()
main = do
  putStrLn "Test Lib3 --------------"
  --quickCheck prop_dynamic_pack_is_fixpoint
  quickCheck prop_dynamic_pack_small_disk


src/Lib3.hs
module Lib3 where

import Control.Monad (liftM2, replicateM)
import Data.Ix
import Data.List (maximumBy, sortBy)
import Text.ParserCombinators.Parsec
  ( Parser,
    anyChar,
    digit,
    eof,
    many,
    many1,
    manyTill,
    newline,
    parse,
    spaces,
  )

-- data Dir = Dir Int String deriving (Show)
data Dir = Dir {dir_size :: Integer, dir_name :: String} deriving (Show, Eq)

--instance Eq Dir where
--(==) a b = (dir_size a == dir_size b) && (dir_name a == dir_name b)

data DirPack = DirPack {pack_size :: Integer, dirs :: [Dir]} deriving (Show)

mediaSize = 700 * 1024 * 1024 :: Integer

greedyPack dirs = foldl maybeAddDir (DirPack 0 []) $ sortBy cmpSize dirs
  where
    cmpSize d1 d2 = compare (dir_size d1) (dir_size d2)

maybeAddDir p d =
  let new_size = pack_size p + dir_size d
      new_dirs = d : dirs p
   in if new_size > mediaSize then p else DirPack new_size new_dirs

parseInput =
  do
    dirs <- many dirAndSize
    eof :: Parser ()
    return dirs

dirAndSize = do
  size <- many1 digit
  spaces
  dir_name <- anyChar `manyTill` newline
  return (Dir (read size) dir_name)

precomputedDisksFor :: [Dir] -> [DirPack]
precomputedDisksFor dirs =
  let precomp = map bestDisk [0 ..]
      bestDisk 0 = DirPack 0 []
      bestDisk limit =
        case [ DirPack (dir_size d + s) (d : ds) | d <- filter (inRange (1, limit) . dir_size) dirs, dir_size d > 0, let (DirPack s ds) = precomp !! fromInteger (limit - dir_size d), d `notElem` ds
             ] of
          [] -> DirPack 0 []
          packs -> maximumBy compSize packs
      compSize a b = compare (pack_size a) (pack_size b)
   in precomp

dynamicPack limit dirs = precomputedDisksFor dirs !! fromInteger limit

someFunc :: IO ()
someFunc = do
  input <- getContents
  putStrLn ("Debug: got input " ++ input)
  let dirs = case parse parseInput "stdin" input of
        Left err -> error $ "Input: \n" ++ show input ++ "\nError:\n" ++ show err
        Right result -> result
  putStrLn "Debug: parsed:"
  print dirs


src/Lib2.hs
module Lib2 where

import Control.Monad (liftM2, replicateM)
import Data.List (sortBy)
import Text.ParserCombinators.Parsec
  ( Parser,
    anyChar,
    digit,
    eof,
    many,
    many1,
    manyTill,
    newline,
    parse,
    spaces,
  )

-- data Dir = Dir Int String deriving (Show)
data Dir = Dir {dir_size :: Int, dir_name :: String} deriving (Show)

instance Eq Dir where
  (==) a b = (dir_size a == dir_size b) && (dir_name a == dir_name b)

data DirPack = DirPack {pack_size :: Int, dirs :: [Dir]} deriving (Show)

media_size = 700 * 1024 * 1024

greedy_pack dirs = foldl maybe_add_dir (DirPack 0 []) $ sortBy cmpSize dirs
  where
    cmpSize d1 d2 = compare (dir_size d1) (dir_size d2)

maybe_add_dir p d =
  let new_size = pack_size p + dir_size d
      new_dirs = d : (dirs p)
   in if new_size > media_size then p else DirPack new_size new_dirs

parseInput =
  do
    dirs <- many dirAndSize
    eof :: Parser ()
    return dirs

dirAndSize = do
  size <- many1 digit
  spaces
  dir_name <- anyChar `manyTill` newline
  return (Dir (read size) dir_name)

-- instance Arbitrary Dir where
--   arbitrary = liftM2 Dir gen_size gen_name
--     where
--       gen_size = do
--         s <- choose (10, 14000)
--         return (s * 1024 * 1024)
--       gen_name = do
--         n <- choose (1, 300)
--         replicateM n (elements "fubar/")

-- prop_greedy_pack_is_fixpoint ds =
--   let pack = greedy_pack ds
--    in pack_size pack == pack_size (greedy_pack (dirs pack))

someFunc :: IO ()
someFunc = do
  input <- getContents
  putStrLn ("Debug: got input " ++ input)
  let dirs = case parse parseInput "stdin" input of
        Left err -> error $ "Input: \n" ++ show input ++ "\nError:\n" ++ show err
        Right result -> result
  putStrLn "Debug: parsed:"
  print dirs

-- main :: IO ()
-- main = do
--   input <- getContents
--   putStrLn ("Debug: got input " ++ input)
--   let dirs = case parse parseInput "stdin" input of
--         Left err -> error $ "Input: \n" ++ show input ++ "\nError:\n" ++ show err
--         Right result -> result
--   putStrLn "Debug: parsed:"
--   print dirs
--   putStrLn "Solution:"
--   print (greedy_pack dirs)
